File: 0091.Decode_Ways.java
//1ms



class Solution {
    public int numDecodings(String s) {
        int n = s.length();
        if(n == 1 && s.charAt(0) != '0') return 1;
        if(s.charAt(0) == '0') return 0;

        int p1 = 0, p2 = 0;

        // for first element
        p2 = s.charAt(n - 1) == '0' ? 0: 1;

        // for second element
        if(s.charAt(n-2) == '0') p1 = 0;
        else if(isDoubleValid(s.charAt(n-2), s.charAt(n-1))) {
            if(s.charAt(n-1) == '0') p1 = 1;
            else p1 = 2;
        } else if(s.charAt(n-1) == '0') p1 = 0;
        else p1 = 1;

        for(int i = n - 3; i >= 0; i--) {
            int temp = p1;
            if(s.charAt(i) != '0') {
                if(isDoubleValid(s.charAt(i), s.charAt(i+1))) p1 += p2;
            } else p1 = 0;
            p2 = temp;
        }

        return p1;
    }

    

    private boolean isDoubleValid(char first, char second) {
        int num = (first - '0') * 10 + (second - '0');
        return num >= 10 && num <= 26;
    }
}

File: 0091.Decode_Ways.py
#0ms


class Solution(object):
    def numDecodings(self, s):
        dp = {}
        def numDecodingsLow(index):
            if index >= len(s):
                return 1
            if s[index] == '0':
                return 0
            if index in dp:
                return dp[index]
            res = numDecodingsLow(index+1)
            if (index+1 < len(s) and 
                ((s[index] == '2' and s[index+1] in "0123456") or
                s[index] == '1')):
                res +=  numDecodingsLow(index+2)
            dp[index] = res
            return res
        return numDecodingsLow(0)
        """
        :type s: str
        :rtype: int
        """
